[SQL] SQL compiler generates mismatched Date/Timestamp window bounds for one-sided temporal filters

## Description

A temporal filter comparing a `DATE` column with an expression derived from `CAST(NOW() AS DATE)` passes SQL compilation, but the generated Rust code does not compile. The window operator uses the `DATE` column as its indexed key while generating a missing boundary as a `TIMESTAMP`, producing incompatible key and boundary types.

## Minimal reproduction

```sql
CREATE TABLE employee (
    id STRING NOT NULL,
    entry_date DATE,
    PRIMARY KEY (id)
) WITH (
    'materialized' = 'true'
);

CREATE MATERIALIZED VIEW label_result AS
SELECT
    id AS entity_id
FROM employee
WHERE entry_date >= CAST(NOW() AS DATE) - INTERVAL '1' YEAR;
```

## Actual behavior

The SQL compiler accepts the program without reporting an error, including when invoked with `sql-to-dbsp -je`.

The generated Rust code then fails during `cargo`/pipeline compilation:

```text
error[E0308]: mismatched types
  --> crates/feldera_pipe_operator_eb055525ea560734/src/lib.rs:72:61
   |
72 |     let operator_eb055525ea560734 = i0.window((true, true), &i1);
   |                                        ------               ^^^ expected `&Stream<ChildCircuit<(), ()>, ...>`, found `&&Stream<ChildCircuit<(), ()>, ...>`
   |                                        |
   |                                        arguments to this method are incorrect
   |
   = note: expected reference `&dbsp::Stream<*, (TypedBox<*, _>, TypedBox<feldera_sqllib::Date, _>)>`
              found reference `&&dbsp::Stream<*, (TypedBox<*, _>, TypedBox<feldera_sqllib::Timestamp, _>)>`
```

The generated operator signature contains mismatched window-bound types:

```rust
i0: &Stream<RootCircuit, IndexedWSet<Date, Tup1<SqlString>>>,

i1: &Stream<
    RootCircuit,
    (
        TypedBox<Date, DynData>,
        TypedBox<Timestamp, DynData>,
    )
>,
```

<img width="1427" height="775" alt="Image" src="https://github.com/user-attachments/assets/22591ce8-ab01-41be-a1dc-7c8826007852" />

The indexed input key is `Date`, but the generated bounds are `(Date, Timestamp)`.

## Expected behavior

The SQL program should compile successfully, with both window boundaries using the indexed input key type:

```text
(Date, Date)
```

Malformed window operators should also be rejected by the SQL compiler before Rust code generation rather than surfacing as a Rust type error.

## Versions tested

The issue was reproduced with:

- Feldera server `0.290.0`
- `sql-to-dbsp` compiler `0.323.0`
- Source based on upstream `main`

The latest compiler inspected still generates the same `eb055525ea560734` operator with `(Date, Timestamp)` bounds.

## Root cause

Temporal filter lowering derives the two window inputs from different type sources.

In `RewriteNow`, the indexed data stream is keyed by the actual type of the common expression. For this query, that type is `Date`:

```text
sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/temporal/RewriteNow.java
```

However, `WindowBounds.makeWindow()` initializes the boundary type with:

```java
DBSPType type = ContainsNow.timestampType();
```

When one side of the window is absent, it generates the default boundary from this hard-coded timestamp type:

```java
if (this.upper != null) {
    upperBound = ...;
} else {
    upperBound = type.to(IsBoundedType.class).getMaxValue();
}
```

For the reproduction:

- the explicit lower bound is a `Date`, derived from `CAST(NOW() AS DATE)`;
- the upper bound is absent;
- the generated default upper bound is the maximum `Timestamp`;
- the resulting control stream therefore has type `(Date, Timestamp)`.

The inverse problem occurs for an upper-only `DATE` filter: the generated bounds have type `(Timestamp, Date)` because the missing lower bound uses the minimum `Timestamp`.

`Date` already implements the bounded-type operations needed to generate its own minimum and maximum values, so the default boundary can use the indexed/common expression type.

`DBSPWindowOperator` currently does not validate the control stream against the indexed input key. Rust code generation consequently emits the malformed operator unchanged. The Rust `window` API requires the indexed key and both boxed boundary values to use the same type `K`, so the mismatch is detected only by `cargo`.

Relevant files:

```text
sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/temporal/WindowBounds.java
sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/temporal/RewriteNow.java
sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPWindowOperator.java
sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
crates/dbsp/src/operator/time_series/window.rs
```

## Scope

The same failure can affect any one-sided temporal filter whose indexed expression is a `DATE`, including strict and non-strict comparisons:

```sql
date_column >  CAST(NOW() AS DATE)
date_column >= CAST(NOW() AS DATE)
date_column <  CAST(NOW() AS DATE)
date_column <= CAST(NOW() AS DATE)
```

Equivalent comparisons with the operands reversed are also affected after comparison normalization.

A true two-sided `DATE` window is not affected by this specific fallback error when both boundaries are supplied explicitly, because neither boundary is generated from the hard-coded timestamp type.

## Workaround

Explicitly convert both the indexed value and computed boundary to `TIMESTAMP`:

```sql
CREATE MATERIALIZED VIEW label_result AS
SELECT
    id AS entity_id
FROM employee
WHERE CAST(entry_date AS TIMESTAMP) >=
      CAST(
          CAST(NOW() AS DATE) - INTERVAL '1' YEAR
          AS TIMESTAMP
      );
```

This produces consistent window types:

```text
(Timestamp, Timestamp)
```

The inner cast to `DATE` preserves the original calendar-date calculation before both operands are converted to `TIMESTAMP` for the window operator.

## Suggested fix

1. Generate missing window boundaries using the non-null type of the common/indexed expression instead of `ContainsNow.timestampType()`.
2. Prefer passing the final indexed key type explicitly to `WindowBounds.makeWindow()` so the data and control streams cannot derive their types independently.
3. Validate in the `DBSPWindowOperator` constructor that:
   - the control stream contains exactly two boxed boundaries;
   - the lower-bound type matches the indexed key type;
   - the upper-bound type matches the indexed key type.
4. Reject malformed operators during Java IR construction, before Rust code generation.

## Regression tests

Please add Rust-compiling regression tests for both open-ended directions:

```sql
-- Lower bound only: previously generated (Date, Timestamp).
WHERE entry_date >= CAST(NOW() AS DATE) - INTERVAL '1' YEAR
```

```sql
-- Upper bound only: previously generated (Timestamp, Date).
WHERE entry_date <= CAST(NOW() AS DATE) + INTERVAL '1' YEAR
```

The tests should compile the generated Rust, not only verify that SQL compilation succeeds. It would also be useful to inspect the resulting `DBSPWindowOperator` and assert that the indexed key, lower boundary, and upper boundary are all the same non-null `Date` type.

A focused negative unit test should additionally verify that constructing a window operator with `Date` indexed keys and `(Date, Timestamp)` or `(Timestamp, Date)` bounds fails immediately.

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPWindowOperator.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPWindowOperator.java
@@ -9,6 +9,9 @@ import org.dbsp.sqlCompiler.compiler.visitors.VisitDecision;
 import org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitVisitor;
 import org.dbsp.sqlCompiler.ir.expression.DBSPExpression;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
+import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeRawTuple;
+import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeIndexedZSet;
+import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeTypedBox;
 import org.dbsp.util.Utilities;
 
 import javax.annotation.Nullable;
@@ -28,8 +31,14 @@ public final class DBSPWindowOperator extends DBSPBinaryOperator implements ICon
     public DBSPWindowOperator(CalciteRelNode node, boolean lowerInclusive, boolean upperInclusive,
             OutputPort data, OutputPort control) {
         super(node, "window", null, data.outputType(), data.isMultiset(), data, control);
-        // Check that the left input and output are indexed ZSets
-        this.getOutputIndexedZSetType();
+        // Check that the left input and output are indexed ZSets.
+        DBSPTypeIndexedZSet indexedType = this.getOutputIndexedZSetType();
+        DBSPType expectedControlType = new DBSPTypeRawTuple(
+                new DBSPTypeTypedBox(indexedType.keyType, false),
+                new DBSPTypeTypedBox(indexedType.keyType, false));
+        Utilities.enforce(control.outputType().sameType(expectedControlType),
+                () -> "Window bounds must have type " + expectedControlType +
+                        ", but have type " + control.outputType());
         this.lowerInclusive = lowerInclusive;
         this.upperInclusive = upperInclusive;
     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/temporal/RewriteNow.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/temporal/RewriteNow.java
@@ -330,32 +330,34 @@ public class RewriteNow extends CircuitCloneVisitor {
         CalciteRelNode relNode = operator.getRelNode();
         DBSPSimpleOperator scalarNow = this.scalarNow();
         WindowBounds bounds = comparisons.getWindowBounds(this.compiler());
-        DBSPClosureExpression makeWindow = bounds.makeWindow();
+        DBSPExpression common = bounds.common();
+        // This exact type is shared by the indexed stream and both window bounds.
+        DBSPType windowKeyType = common.getType().withMayBeNull(false);
+        DBSPClosureExpression makeWindow = bounds.makeWindow(windowKeyType);
         DBSPSimpleOperator windowBounds = new DBSPApplyOperator(operator.getRelNode(),
                 makeWindow, scalarNow.outputPort(), null);
         this.addOperator(windowBounds);
 
-        // Filter the null timestamps away, they won't be selected anyway,
-        // but window needs non-nullable values
+        // Filter null keys away; they would not be selected anyway, and the
+        // window operator requires non-nullable keys.
         DBSPTypeTupleBase inputType = source.getOutputZSetElementType().to(DBSPTypeTupleBase.class);
-        DBSPType commonType = bounds.common().getType();
         DBSPParameter param = comparisons.getParameter();
-        if (bounds.common().getType().mayBeNull) {
+        if (common.getType().mayBeNull) {
             DBSPClosureExpression nonNull =
-                    bounds.common().is_null().not().closure(param);
+                    common.is_null().not().closure(param);
             DBSPFilterOperator filter = new DBSPFilterOperator(relNode, nonNull, source.outputPort());
             this.addOperator(filter);
             source = filter;
         }
 
-        // Index input by timestamp
+        // Index input by the temporal key.
         DBSPClosureExpression indexFunction =
                 new DBSPRawTupleExpression(
-                        bounds.common().cast(bounds.common().getNode(), commonType.withMayBeNull(false),
+                        common.cast(common.getNode(), windowKeyType,
                                 DBSPCastExpression.CastType.SqlUnsafe),
                         param.asVariable().deref().applyClone()).closure(param);
         DBSPTypeIndexedZSet ix = new DBSPTypeIndexedZSet(operator.getRelNode(),
-                commonType.withMayBeNull(false),
+                windowKeyType,
                 inputType);
         DBSPMapIndexOperator index = new DBSPMapIndexOperator(operator.getRelNode(),
                 indexFunction, ix, operator.isMultiset, source.outputPort());

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/temporal/WindowBounds.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/temporal/WindowBounds.java
@@ -2,6 +2,7 @@ package org.dbsp.sqlCompiler.compiler.visitors.outer.temporal;
 
 import org.dbsp.sqlCompiler.compiler.DBSPCompiler;
 import org.dbsp.sqlCompiler.compiler.frontend.calciteObject.CalciteObject;
+import org.dbsp.sqlCompiler.ir.expression.DBSPCastExpression;
 import org.dbsp.sqlCompiler.ir.expression.DBSPClosureExpression;
 import org.dbsp.sqlCompiler.ir.expression.DBSPExpression;
 import org.dbsp.sqlCompiler.ir.expression.DBSPRawTupleExpression;
@@ -18,25 +19,28 @@ record WindowBounds(
         @Nullable WindowBound lower,
         @Nullable WindowBound upper,
         DBSPExpression common) {
-    DBSPClosureExpression makeWindow() {
-        DBSPType type = ContainsNow.timestampType();
-        // The input has type Tup1<Timestamp>
-        DBSPVariablePath var = new DBSPTypeTuple(type).ref().var();
-        RewriteNow.RewriteNowExpression rn = new RewriteNow.RewriteNowExpression(this.compiler(), var.deref().field(0));
+    DBSPClosureExpression makeWindow(DBSPType windowKeyType) {
+        // The input has type Tup1<Timestamp>, since it is produced by scalarNow().
+        DBSPType timestampType = ContainsNow.timestampType();
+        DBSPVariablePath var = new DBSPTypeTuple(timestampType).ref().var();
+        RewriteNow.RewriteNowExpression rn = new RewriteNow.RewriteNowExpression(
+                this.compiler(), var.deref().field(0));
         final DBSPExpression lowerBound, upperBound;
         CalciteObject node = CalciteObject.EMPTY;
         if (this.lower != null) {
-            lowerBound = rn.apply(this.lower().expression()).to(DBSPExpression.class);
+            lowerBound = rn.apply(this.lower().expression()).to(DBSPExpression.class)
+                    .nullabilityCast(windowKeyType, DBSPCastExpression.CastType.SqlUnsafe);
             node = this.lower.expression().getNode();
         } else {
-            lowerBound = type.to(IsBoundedType.class).getMinValue();
+            lowerBound = windowKeyType.to(IsBoundedType.class).getMinValue();
         }
         if (this.upper != null) {
-            upperBound = rn.apply(this.upper.expression()).to(DBSPExpression.class);
+            upperBound = rn.apply(this.upper.expression()).to(DBSPExpression.class)
+                    .nullabilityCast(windowKeyType, DBSPCastExpression.CastType.SqlUnsafe);
             if (!node.getPositionRange().isValid())
                 node = this.upper.expression().getNode();
         } else {
-            upperBound = type.to(IsBoundedType.class).getMaxValue();
+            upperBound = windowKeyType.to(IsBoundedType.class).getMaxValue();
         }
         return new DBSPRawTupleExpression(node,
                 DBSPTypeTypedBox.wrapTypedBox(lowerBound, false),

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/circuit/operator/DBSPWindowOperatorTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/circuit/operator/DBSPWindowOperatorTests.java
@@ -1,0 +1,95 @@
+package org.dbsp.sqlCompiler.circuit.operator;
+
+import org.dbsp.sqlCompiler.compiler.errors.InternalCompilerError;
+import org.dbsp.sqlCompiler.circuit.OutputPort;
+import org.dbsp.sqlCompiler.compiler.frontend.calciteObject.CalciteEmptyRel;
+import org.dbsp.sqlCompiler.compiler.frontend.calciteObject.CalciteObject;
+import org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitVisitor;
+import org.dbsp.sqlCompiler.ir.expression.DBSPExpression;
+import org.dbsp.sqlCompiler.ir.expression.DBSPIndexedZSetExpression;
+import org.dbsp.sqlCompiler.ir.expression.DBSPRawTupleExpression;
+import org.dbsp.sqlCompiler.ir.expression.literal.DBSPDateLiteral;
+import org.dbsp.sqlCompiler.ir.expression.literal.DBSPTimestampLiteral;
+import org.dbsp.sqlCompiler.ir.type.DBSPType;
+import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeTuple;
+import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeDate;
+import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeIndexedZSet;
+import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeTypedBox;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.List;
+
+public class DBSPWindowOperatorTests {
+    private static DBSPConstantOperator dataInput() {
+        DBSPTypeIndexedZSet type = new DBSPTypeIndexedZSet(
+                CalciteObject.EMPTY, DBSPTypeDate.INSTANCE, DBSPTypeTuple.EMPTY);
+        return new DBSPConstantOperator(CalciteEmptyRel.INSTANCE,
+                new DBSPIndexedZSetExpression(CalciteObject.EMPTY, type), false);
+    }
+
+    private static OutputPort controlInput(
+            DBSPExpression lower, DBSPExpression upper) {
+        DBSPRawTupleExpression bounds = new DBSPRawTupleExpression(
+                DBSPTypeTypedBox.wrapTypedBox(lower, false),
+                DBSPTypeTypedBox.wrapTypedBox(upper, false));
+        DBSPApplyOperator apply = new DBSPApplyOperator(
+                CalciteEmptyRel.INSTANCE,
+                bounds.closure(DBSPTypeTuple.EMPTY.ref().var()),
+                bounds.getType(),
+                new TestScalarOperator().outputPort(),
+                null);
+        return apply.outputPort();
+    }
+
+    private static DBSPWindowOperator window(
+            DBSPExpression lower, DBSPExpression upper) {
+        return new DBSPWindowOperator(CalciteEmptyRel.INSTANCE, true, true,
+                dataInput().outputPort(), controlInput(lower, upper));
+    }
+
+    private static final class TestScalarOperator extends DBSPSimpleOperator {
+        TestScalarOperator() {
+            super(CalciteEmptyRel.INSTANCE, "apply_test_source", null,
+                    DBSPTypeTuple.EMPTY, false);
+        }
+
+        @Override
+        public DBSPSimpleOperator with(
+                DBSPExpression function,
+                DBSPType outputType,
+                List<OutputPort> newInputs,
+                boolean force) {
+            return this;
+        }
+
+        @Override
+        public void accept(CircuitVisitor visitor) {
+            visitor.push(this);
+            visitor.pop(this);
+        }
+    }
+
+    @Test
+    public void acceptsMatchingBoundTypes() {
+        DBSPWindowOperator window = window(
+                new DBSPDateLiteral("2025-01-01"),
+                new DBSPDateLiteral("2025-12-31"));
+        Assert.assertTrue(window.left().getOutputIndexedZSetType().keyType
+                .sameType(DBSPTypeDate.INSTANCE));
+    }
+
+    @Test
+    public void rejectsMismatchedUpperBoundType() {
+        Assert.assertThrows(InternalCompilerError.class, () -> window(
+                new DBSPDateLiteral("2025-01-01"),
+                new DBSPTimestampLiteral("2025-12-31 00:00:00", false)));
+    }
+
+    @Test
+    public void rejectsMismatchedLowerBoundType() {
+        Assert.assertThrows(InternalCompilerError.class, () -> window(
+                new DBSPTimestampLiteral("2025-01-01 00:00:00", false),
+                new DBSPDateLiteral("2025-12-31")));
+    }
+}

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegressionTests.java
@@ -30,8 +30,11 @@ import org.dbsp.sqlCompiler.ir.expression.literal.DBSPStringLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPU64Literal;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeFunction;
+import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeRawTuple;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeAny;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBool;
+import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeDate;
+import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeTypedBox;
 import org.junit.Assert;
 import org.dbsp.sqlCompiler.compiler.sql.tools.SqlIoTest;
 import org.junit.Test;
@@ -837,6 +840,47 @@ public class IncrementalRegressionTests extends SqlIoTest {
                         window.left().getOutputIndexedZSetType().elementType.getToplevelFieldCount());
             }
         });
+    }
+
+    private void testDateTemporalFilter(String comparison) {
+        var ccs = this.getCCS("""
+                CREATE TABLE employee (
+                    id STRING NOT NULL,
+                    entry_date DATE,
+                    PRIMARY KEY (id)
+                );
+                CREATE VIEW label_result AS
+                SELECT id AS entity_id
+                FROM employee
+                WHERE entry_date %s;
+                """.formatted(comparison));
+        ccs.visit(new CircuitVisitor(ccs.compiler) {
+            int windows;
+
+            @Override public void postorder(DBSPWindowOperator window) {
+                this.windows++;
+                DBSPType keyType = window.left().getOutputIndexedZSetType().keyType;
+                Assert.assertTrue(keyType.sameType(DBSPTypeDate.INSTANCE));
+                DBSPType expectedControlType = new DBSPTypeRawTuple(
+                        new DBSPTypeTypedBox(keyType, false),
+                        new DBSPTypeTypedBox(keyType, false));
+                Assert.assertTrue(window.right().outputType().sameType(expectedControlType));
+            }
+
+            @Override public void endVisit() {
+                Assert.assertEquals(1, this.windows);
+            }
+        });
+    }
+
+    @Test
+    public void dateTemporalFilterWithLowerBound() {
+        this.testDateTemporalFilter(">= CAST(NOW() AS DATE) - INTERVAL '1' YEAR");
+    }
+
+    @Test
+    public void dateTemporalFilterWithUpperBound() {
+        this.testDateTemporalFilter("<= CAST(NOW() AS DATE) + INTERVAL '1' YEAR");
     }
 
     @Test

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegressionTests.java
@@ -29,11 +29,13 @@ import org.dbsp.sqlCompiler.ir.expression.literal.DBSPI32Literal;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPStringLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPU64Literal;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
+import org.dbsp.sqlCompiler.ir.type.DBSPTypeCode;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeFunction;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeRawTuple;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeAny;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBool;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeDate;
+import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeInteger;
 import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeTypedBox;
 import org.junit.Assert;
 import org.dbsp.sqlCompiler.compiler.sql.tools.SqlIoTest;
@@ -881,6 +883,48 @@ public class IncrementalRegressionTests extends SqlIoTest {
     @Test
     public void dateTemporalFilterWithUpperBound() {
         this.testDateTemporalFilter("<= CAST(NOW() AS DATE) + INTERVAL '1' YEAR");
+    }
+
+    private void testIntegerTemporalFilter(String comparison) {
+        var ccs = this.getCCS("""
+                CREATE TABLE events (
+                    id STRING NOT NULL,
+                    event_epoch BIGINT NOT NULL,
+                    PRIMARY KEY (id)
+                );
+                CREATE VIEW recent_events AS
+                SELECT id AS entity_id
+                FROM events
+                WHERE event_epoch %s;
+                """.formatted(comparison));
+        ccs.visit(new CircuitVisitor(ccs.compiler) {
+            int windows;
+
+            @Override public void postorder(DBSPWindowOperator window) {
+                this.windows++;
+                DBSPType keyType = window.left().getOutputIndexedZSetType().keyType;
+                Assert.assertTrue(keyType.sameType(
+                        DBSPTypeInteger.getType(CalciteObject.EMPTY, DBSPTypeCode.INT64, false)));
+                DBSPType expectedControlType = new DBSPTypeRawTuple(
+                        new DBSPTypeTypedBox(keyType, false),
+                        new DBSPTypeTypedBox(keyType, false));
+                Assert.assertTrue(window.right().outputType().sameType(expectedControlType));
+            }
+
+            @Override public void endVisit() {
+                Assert.assertEquals(1, this.windows);
+            }
+        });
+    }
+
+    @Test
+    public void integerTemporalFilterWithLowerBound() {
+        this.testIntegerTemporalFilter(">= CAST(NOW() AS BIGINT) - 86400");
+    }
+
+    @Test
+    public void integerTemporalFilterWithUpperBound() {
+        this.testIntegerTemporalFilter("<= CAST(NOW() AS BIGINT) + 86400");
     }
 
     @Test

